### PR TITLE
fix(core menu): Menuize the account routes so they can be added to by…

### PR DESCRIPTION
**Pull request by trainerbill**
… submodules

Removed hard coded account menus from header view and added them to a menu file.  Will allow submodules to add to the menu if they want to add user specific routes.

### CODE REVIEW

File: modules/core/client/views/header.client.view.html [hunk 1]
```diff
@@ -36,17 +36,8 @@
...
-          </li>
-          <li ui-sref-active="active">
-            <a ui-sref="settings.picture">Change Profile Picture</a>
-          </li>
-          <li ui-sref-active="active" ng-show="authentication.user.provider === 'local'">
-            <a ui-sref="settings.password">Change Password</a>
-          </li>
-          <li ui-sref-active="active">
-            <a ui-sref="settings.accounts">Manage Social Accounts</a>
+          <li ui-sref-active="active" ng-repeat="item in accountMenu.items[0].items">
```
Comment (mleanos) [inline modules/core/client/views/header.client.view.html]: Can this be changed, so we don't have to rely on an index? Just set the object that you want to iterate through, to your $scope.accountMenu object.

```javascript
// in your controller
$scope.accountMenu = Menus.getMenu('account')[0];
```

File: modules/core/client/config/core.client.menus.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      roles: ['user']
+    });
+
+    Menus.addMenuItem('account', {
+      title: '',
+      state: 'settings',
+      type: 'dropdown',
+      roles: ['user']
+    });
+
```
Comment (mleanos) [inline modules/core/client/config/core.client.menus.js]: I've always been confused by the implementation of the Menus service. It never seems intuitive to me.

When would we have multiple *Menu Items* defined for a Menu?
One menu defined with?
&nbsp; Item1
&nbsp; &nbsp; SubItem1
&nbsp; &nbsp; SubItem2
&nbsp; Item2
&nbsp; &nbsp; SubItemA
&nbsp; &nbsp; SubItemB

Makes sense to me for the Account menu:
Edit Profile
Edit Profile Pic
Manage Social Account

It seems that the options defined in the MenuItem should be defined at the Menu level. Maybe I'm looking at all this wrong? Like I said, every time I look at this service it confuses me :)

**Comment by mleanos** [comment]
I think this is a great enhancement. I left a couple of comments. Basically, I've never liked how the Menus service was implemented. Maybe we want to take the time to refactor this service now?

File: modules/core/client/config/core.client.menus.js (hunk 2 above)
Comment (trainerbill) [inline modules/core/client/config/core.client.menus.js]: Agreed the menu structure is odd.  Took me a while to find out that I had to do an addMenu.  Seems like too many levels.  I checked out the Bootstrap 4 docs and it looks fairly compatible, but I would wait until it drops before rewriting the Menu service:

http://v4-alpha.getbootstrap.com/components/navs/

**Comment by trainerbill** [comment]
@mleanos I made the fixes and rebased.  Can we get this pushed in since it is such a small update?  I would rather wait on a full Menu refactor till Bootstrap 4.

File: modules/core/client/config/core.client.menus.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      title: 'Change Password',
+      state: 'settings.password'
+    });
+
+    Menus.addSubMenuItem('account', 'settings', {
+      title: 'Manage Social Accounts',
+      state: 'settings.accounts'
+    });
+
+
```
Comment (codydaig) [inline modules/core/client/config/core.client.menus.js]: Excess newlines.

**Comment by codydaig** [comment]
Other than the group of excess new lines, LGTM.

**Comment by trainerbill** [comment]
@codydaig new lines fixed.  I say merge it.

**Comment by codydaig** [comment]
I would like to get @ilanbiala's or @lirantal's thoughts before I merge. Make sure there's not something else that I am overlooking.

File: modules/core/client/config/core.client.menus.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+(function () {
+  'use strict';
+
+  // Configuring the Fishing module
```
Comment (ilanbiala) [inline modules/core/client/config/core.client.menus.js]: Fishing?

File: modules/core/client/views/header.client.view.html [hunk 5]
```diff
@@ -36,17 +36,8 @@
...
-          <li ui-sref-active="active">
-            <a ui-sref="settings.picture">Change Profile Picture</a>
-          </li>
-          <li ui-sref-active="active" ng-show="authentication.user.provider === 'local'">
-            <a ui-sref="settings.password">Change Password</a>
-          </li>
-          <li ui-sref-active="active">
-            <a ui-sref="settings.accounts">Manage Social Accounts</a>
+          <li ui-sref-active="active" ng-repeat="item in accountMenu.items">
+            <a ui-sref="{{item.state}}" ng-bind="item.title"></a>
```
Comment (ilanbiala) [inline modules/core/client/views/header.client.view.html]: Why ng-bind and not {{}}?

File: modules/core/client/views/header.client.view.html [hunk 6]
```diff
@@ -36,17 +36,8 @@
...
-          </li>
-          <li ui-sref-active="active">
-            <a ui-sref="settings.picture">Change Profile Picture</a>
-          </li>
-          <li ui-sref-active="active" ng-show="authentication.user.provider === 'local'">
-            <a ui-sref="settings.password">Change Password</a>
-          </li>
-          <li ui-sref-active="active">
-            <a ui-sref="settings.accounts">Manage Social Accounts</a>
+          <li ui-sref-active="active" ng-repeat="item in accountMenu.items">
```
Comment (ilanbiala) [inline modules/core/client/views/header.client.view.html]: item should be menu because semantically it's more logical.

Comment (trainerbill) [inline modules/core/client/views/header.client.view.html]: I just used the same that was there for the rest of the menus:

https://github.com/trainerbill/mean/blob/MenusFixes/modules/core/client/views/header.client.view.html#L13

File: modules/core/client/views/header.client.view.html (hunk 5 above)
Comment (trainerbill) [inline modules/core/client/views/header.client.view.html]: Same syntax that is currently there...

**Comment by trainerbill** [comment]
Removed the comment.  LGTM?

**Comment by ilanbiala** [comment]
@codydaig LGTM.

**Comment by mleanos** [comment]
LGTM. I pulled it down and tested.

**Comment by ilanbiala** [comment]
@trainerbill take a look at CI, something seems to have failed.

**Comment by trainerbill** [comment]
@ilanbiala I noticed that as well but they are all in the Chat module which I did not touch.  I am up to date on master and when i checkout upstream/master and do grunt test I am getting the same errors.  Can someone test it out locally?  I think some other PR may have broken something.

**Comment by trainerbill** [comment]
Ok, just to make sure I wasn't crazy, I cloned a fresh mean, npm install using 0.12, and grunt test and the same tests fail. @ilanbiala can you clone down a fresh copy and see if the tests are passing?  Or anyone else.  My dev env is pretty complex, but I don't think its an issue with the PR or my env.

**Comment by mleanos** [comment]
@trainerbill @ilanbiala The tests are passing fine for me locally, with an up-to-date version of master.

Windows 7
NPM: 2.11.3
Node: v0.12.7

Are you having the same behavior with running `gulp test:client`?

**Comment by mleanos** [comment]
@trainerbill It looks like there's an issue with the Socket.IO module. When I re-installed Socket.IO, it resolved to installing v1.4.5. I then ran `grunt test:client` again, and got an error thrown complaining about the Socket.

Can you try re-installing Socket.IO, and setting the dependency to `"socket.io": "~1.3.7"`? If that resolves the issue for you then we'll know it's with Socket.IO, and can look at fixing it in a separate PR.

**Comment by trainerbill** [comment]
@mleanos I downgraded to 1.3.7 and am still getting the same errors on this PR, and a fresh clone of mean master, using node 0.12.9 and 4.2.6.  I have run npm install/rebuild more times than I can count.  I am on a linux OS, which Travis CI is as well.  Maybe that is the issue?

**Comment by trainerbill** [comment]
@mleanos The issue is with phantomjs.  changing socketio to ~1.3.7 and phantomjs to ~1.9.0 works

**Comment by trainerbill** [comment]
Since this PR is LGTM, I am going to update phantomjs and socketio to use minors ~.  I had @codydaig rerun a passing PR and it is now failing.

https://travis-ci.org/meanjs/mean/jobs/102975444

**Comment by trainerbill** [comment]
@codydaig @ilanbiala This PR now fixes Travis build errors by requiring PhantomJS ~1.9.0 and SocketIO ~1.3.7.  Can we merge?

**Comment by codydaig** [comment]
@ilanbiala This PR passes with the updated dependencies. Merging in.

**Pull request merged**